plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java

```diff
@@ -454,4 +430,44 @@ public static List<ParquetReaderColumn> createParquetReaderColumns(List<HiveColu
                 })
                 .collect(toImmutableList());
     }
+
+    private static Optional<Type> getBaseColumnType(HiveColumnHandle column, MessageType messageType, boolean useParquetColumnNames)
+    {
+        if (useParquetColumnNames) {
+            return Optional.ofNullable(getParquetTypeByName(column.getBaseColumnName(), messageType));
+        }
+        if (column.getBaseHiveColumnIndex() < messageType.getFieldCount()) {
+            return Optional.of(messageType.getType(column.getBaseHiveColumnIndex()));
+        }
+
+        return Optional.empty();
+    }
+
+    /**
+     * Dereferencing base parquet type based on projection info's dereference names.
+     * For example, when dereferencing baseType(level1Field0, level1Field1, Level1Field2(Level2Field0, Level2Field1))
+     * with a projection info's dereferenceNames list as (basetype, Level1Field2, Level2Field1).
+     * It would return a list of parquet types in the order of (level1Field2, Level2Field1).*
+     *
+     * @return child fields on each level of dereferencing. Return emptyList when failed to do lookup.
+     */
+    private static List<Type> dereferenceSubFieldTypes(GroupType baseType, HiveColumnProjectionInfo projectionInfo)
+    {
+        checkArgument(baseType != null, "base type cannot be null when dereferencing");
+        checkArgument(projectionInfo != null, "hive column projection info cannot be null when doing dereferencing");
+
+        ImmutableList.Builder<Type> typeBuilder = ImmutableList.builder();
+        Type parentType = baseType;
+
+        for (String name : projectionInfo.getDereferenceNames()) {
+            Type childType = getParquetTypeByName(name, parentType.asGroupType());
```

Comment:
Should this use an index lookup if `useColumnNames` is set to false?

Reply:
Thanks, I also have the same question. 

Since this PR is related to predicate push down. I would like treat `useColumnNames` in a separate task. And only extract out this common method and reuse the previous logic at the current PR. 

Changing how Trino look up subfields can cause backward incompatible change. How do you think?

Furthermore, since it is essentially looking up nested field, if we indeed want to apply this config, maybe we should rename the config from `useColumnNames` to `useFieldNames` to cover both lookup by name/index at the column level and at the subfield level?

Reply:
From the definition of the property:

```
 booleanProperty(
                        PARQUET_USE_COLUMN_NAME,
                        "Parquet: Access Parquet columns using names from the file",
                        hiveConfig.isUseParquetColumnNames(),
                        false)
```
useColumnNames seems only enable/disable accessing parquet columns using names.

Reply:
let me know if we should honor PARQUET_USE_COLUMN_NAME when dereferencing subfields.